### PR TITLE
[SPARK-19426][SQL] Custom coalesce for Dataset

### PR DESCRIPTION
**Pull request by mariusvniekerk**
## What changes were proposed in this pull request?

This adds support for using the PartitionCoalescer features added in #11865 (SPARK-14042) to the Dataset API

## How was this patch tested?

Manual tests

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 1]
```diff
@@ -823,6 +825,17 @@ case class Repartition(numPartitions: Int, shuffle: Boolean, child: LogicalPlan)
 }
 
 /**
+  * Returns a new RDD that has exactly `numPartitions` partitions.
+  */
+case class CoalesceLogical(numPartitions: Int, partitionCoalescer: Option[PartitionCoalescer],
+                    child: LogicalPlan)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Could you follow the styles documented in https://github.com/databricks/scala-style-guide?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 2]
```diff
@@ -823,6 +825,17 @@ case class Repartition(numPartitions: Int, shuffle: Boolean, child: LogicalPlan)
 }
 
 /**
+  * Returns a new RDD that has exactly `numPartitions` partitions.
+  */
+case class CoalesceLogical(numPartitions: Int, partitionCoalescer: Option[PartitionCoalescer],
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: `CoalesceLogical ` -> `Coalesce`?

**Comment by gatorsmile** [comment]
Could you please also add a few test cases?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 2 above)
Comment (mariusvniekerk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Main reason is there was already a Coalesce expression class

**Comment by felixcheung** [comment]
I'd second that. I'd be interested to know if this implementation changes behavior for `coalesce`

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 3]
```diff
@@ -2437,9 +2435,12 @@ class Dataset[T] private[sql](
...
    */
-  def coalesce(numPartitions: Int): Dataset[T] = withTypedPlan {
-    Repartition(numPartitions, shuffle = false, logicalPlan)
-  }
+  def coalesce(numPartitions: Int, partitionCoalescer: Option[PartitionCoalescer]): Dataset[T] =
+    withTypedPlan {
+      CoalesceLogical(numPartitions, partitionCoalescer, logicalPlan)
+    }
+
+  def coalesce(numPartitions: Int): Dataset[T] = coalesce(numPartitions, None)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Please also add the function description, like what we did in the other functions in Dataset.scala?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala [hunk 4]
```diff
@@ -497,7 +496,9 @@ case class UnionExec(children: Seq[SparkPlan]) extends SparkPlan {
  * if you go from 1000 partitions to 100 partitions, there will not be a shuffle, instead each of
  * the 100 new partitions will claim 10 of the current partitions.
  */
-case class CoalesceExec(numPartitions: Int, child: SparkPlan) extends UnaryExecNode {
+case class CoalesceExec(numPartitions: Int, child: SparkPlan,
+                        partitionCoalescer: Option[PartitionCoalescer]
+                       ) extends UnaryExecNode {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: The same indent issue here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 5]
```diff
@@ -17,7 +17,9 @@
 
 package org.apache.spark.sql.catalyst.plans.logical
 
+import org.apache.spark.rdd.PartitionCoalescer
 import org.apache.spark.sql.catalyst.{CatalystConf, TableIdentifier}
+import scala.collection.mutable.ArrayBuffer
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Useless?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 6]
```diff
@@ -823,6 +825,17 @@ case class Repartition(numPartitions: Int, shuffle: Boolean, child: LogicalPlan)
 }
 
 /**
+ * Returns a new RDD that has exactly `numPartitions` partitions.
+ */
+case class CoalesceLogical(numPartitions: Int, partitionCoalescer: Option[PartitionCoalescer],
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: The name still looks inconsistent with the others. How about `PartitionCoalesce`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 7]
```diff
@@ -823,6 +825,17 @@ case class Repartition(numPartitions: Int, shuffle: Boolean, child: LogicalPlan)
 }
 
 /**
+ * Returns a new RDD that has exactly `numPartitions` partitions.
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: This description is not right.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 8]
```diff
@@ -823,6 +825,17 @@ case class Repartition(numPartitions: Int, shuffle: Boolean, child: LogicalPlan)
...
+ */
+case class CoalesceLogical(numPartitions: Int, partitionCoalescer: Option[PartitionCoalescer],
+                    child: LogicalPlan)
+  extends UnaryNode {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]:
```Scala
case class PartitionCoalesce(
    numPartitions: Int,
    partitionCoalescer: Option[PartitionCoalescer],
    child: LogicalPlan) extends UnaryNode {
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala [hunk 9]
```diff
@@ -19,9 +19,8 @@ package org.apache.spark.sql.execution
 
 import scala.concurrent.{ExecutionContext, Future}
 import scala.concurrent.duration.Duration
-
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: Add it back?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala (hunk 4 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]:
```
case class CoalesceExec(
    numPartitions: Int,
    child: SparkPlan,
    partitionCoalescer: Option[PartitionCoalescer]) extends UnaryExecNode {
```

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 10]
```diff
@@ -117,6 +134,34 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
...
 
+  test("coalesce, custom") {
+
+    val maxSplitSize = 512
+    // Similar to the implementation of `test("custom RDD coalescer")` from [[RDDSuite]] we first
+    // write out to disk, to ensure that our splits are in fact [[FileSplit]] instances.
+    val data = (1 to 1000).map(i => ClassData(i.toString, i))
+    data.toDS().repartition(10).write.format("csv").save(path.toString)
+
+    val ds = spark.read.format("csv").load(path.toString).as[ClassData]
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]:
```
cannot resolve '`a`' given input columns: [_c0, _c1];
```

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 11]
```diff
@@ -117,6 +134,34 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
       data: _*)
   }
 
+  test("coalesce, custom") {
+
+    val maxSplitSize = 512
+    // Similar to the implementation of `test("custom RDD coalescer")` from [[RDDSuite]] we first
+    // write out to disk, to ensure that our splits are in fact [[FileSplit]] instances.
+    val data = (1 to 1000).map(i => ClassData(i.toString, i))
+    data.toDS().repartition(10).write.format("csv").save(path.toString)
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: use `WithPath` to generate the path?

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 12]
```diff
@@ -17,24 +17,41 @@
...
+
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    path = Utils.createTempDir()
+    path.delete()
+  }
+
+  after {
+    Utils.deleteRecursively(path)
+  }
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: No need to do it, if you use `withTempPath`. [This](https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala#L247-L265) is an example

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 13]
```diff
@@ -117,6 +134,34 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
...
+      var splitSizeSum = 0L
+      partition.asInstanceOf[CoalescedRDDPartition].parents.foreach(partition => {
+        val split = partition.asInstanceOf[HadoopPartition].inputSplit.value.asInstanceOf[FileSplit]
+        splitSizeSum += split.getLength
+        totalPartitionCount += 1
+      })
+      assert(splitSizeSum <= maxSplitSize)
+    })
+    assert(totalPartitionCount == 10)
+
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: Nit: Remove this empty line.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 6 above)
Comment (mariusvniekerk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: that sounds good

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala (hunk 12 above)
Comment (mariusvniekerk) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: ah thanks.  I looked at the writer tests

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala (hunk 10 above)
Comment (mariusvniekerk) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: Oh right csv doesn't do headers.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala (hunk 4 above)
Comment (mariusvniekerk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: Do you guys have a .scalafmt.conf that applies all of this?  that should make things cleaner.

**Comment by mariusvniekerk** [comment]
@felixcheung This does not touch any of the coalesce internals.  Only allows setting a partitionCoalescer  similar to what is already available in rdd.coalesce

**Comment by gatorsmile** [comment]
Sorry for the late reply. @mariusvniekerk 

Could you please update the PR?

**Comment by gatorsmile** [comment]
ok to test

**Comment by mariusvniekerk** [comment]
Let me rebase this.  I don't currently have a clean way of testing this on Windows

**Comment by gatorsmile** [comment]
Could you run the following four commands to do a local test in your local environment?
```
dev/lint-scala
build/sbt -Phive hive/test
build/sbt sql/test
build/sbt catalyst/test
```

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 14]
```diff
@@ -2603,12 +2603,27 @@ class Dataset[T] private[sql](
    * current upstream partitions will be executed in parallel (per whatever
    * the current partitioning is).
    *
+   * A [[PartitionCoalescer]] can also be supplied allowing the behavior of the partitioning to be
+   * customized similar to [[RDD.coalesce]].
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: I think it should be `[[org.apache.spark.rdd.RDD##coalesce]]`.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 15]
```diff
@@ -2603,12 +2603,27 @@ class Dataset[T] private[sql](
    * current upstream partitions will be executed in parallel (per whatever
    * the current partitioning is).
    *
+   * A [[PartitionCoalescer]] can also be supplied allowing the behavior of the partitioning to be
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Sounds this trait is unable to be generated as is in Java. Simply wrapping `` `...` `` would be fine.

**Review by HyukjinKwon — COMMENTED** [review]
Actual javadoc errors are as below:

```
[error] /home/jenkins/workspace/SparkPullRequestBuilder/sql/core/target/java/org/apache/spark/sql/Dataset.java:2222: error: reference not found
[error]    * A {@link PartitionCoalescer} can also be supplied allowing the behavior of the partitioning to be
[error]               ^
[error] /home/jenkins/workspace/SparkPullRequestBuilder/sql/core/target/java/org/apache/spark/sql/Dataset.java:2223: error: reference not found
[error]    * customized similar to {@link RDD.coalesce}.
[error]                   
```

**Comment by HyukjinKwon** [comment]
Hi @mariusvniekerk, would you be able to fix the javadoc errors?

**Comment by gatorsmile** [comment]
cc @maropu Do you want to take this over?

**Comment by maropu** [comment]
@gatorsmile Sure! I'll do, Thanks!

**Pull request closed without merging**